datasources-agroal/src/main/java/org/wildfly/extension/datasources/agroal/AbstractDataSourceDefinition.java

```diff
@@ -0,0 +1,247 @@
+import static org.jboss.as.controller.SimpleAttributeDefinitionBuilder.create;
+
+/**
+ * Common Definition for the datasource resource
+ *
+ * @author <a href="lbarreiro@redhat.com">Luis Barreiro</a>
+ */
+abstract class AbstractDataSourceDefinition extends PersistentResourceDefinition {
+
+    private static final String DATA_SOURCE_CAPABILITY_NAME = "org.wildfly.data-source";
+
+    public static final RuntimeCapability<Void> DATA_SOURCE_CAPABILITY = RuntimeCapability.Builder.of(DATA_SOURCE_CAPABILITY_NAME, true, DataSource.class).build();
+
+    static final SimpleAttributeDefinition JNDI_NAME_ATTRIBUTE = create("jndi-name", ModelType.STRING)
+            .setAllowExpression(true)
+            .setRestartAllServices()
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final SimpleAttributeDefinition STATISTICS_ENABLED_ATTRIBUTE = create("statistics-enabled", ModelType.BOOLEAN)
+            .setAllowExpression(true)
+            .setDefaultValue(new ModelNode(false))
+            .setRequired(false)
+            .build();
+
+    // --- connection-factory attributes //
+
+    static final SimpleAttributeDefinition DRIVER_ATTRIBUTE = create("driver", ModelType.STRING)
+            .setRequires()
+            .setRestartAllServices()
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final SimpleAttributeDefinition URL_ATTRIBUTE = create("url", ModelType.STRING)
+            .setAllowExpression(true)
+            .setRestartAllServices()
+            .setRequired(false)
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final SimpleAttributeDefinition TRANSACTION_ISOLATION_ATTRIBUTE = create("transaction-isolation", ModelType.STRING)
+            .setAllowExpression(true)
+            .setAllowedValues(NONE.name(), READ_UNCOMMITTED.name(), READ_COMMITTED.name(), REPEATABLE_READ.name(), SERIALIZABLE.name())
+            .setRequired(false)
+            .setRestartAllServices()
+            .setValidator(EnumValidator.create(AgroalConnectionFactoryConfiguration.TransactionIsolation.class, EnumSet.allOf(AgroalConnectionFactoryConfiguration.TransactionIsolation.class)))
+            .build();
+
+    static final SimpleAttributeDefinition NEW_CONNECTION_SQL_ATTRIBUTE = create("new-connection-sql", ModelType.STRING)
+            .setAllowExpression(true)
+            .setRequired(false)
+            .setRestartAllServices()
+            .build();
+
+    static final SimpleAttributeDefinition USERNAME_ATTRIBUTE = create("username", ModelType.STRING)
+            .setAttributeGroup("security")
+            .setAllowExpression(true)
+            .setRequired(false)
+            .setRestartAllServices()
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final SimpleAttributeDefinition PASSWORD_ATTRIBUTE = create("password", ModelType.STRING)
```

Comment:
I wonder if it shouldn't support also credential-reference, users don't like clear text password in configuration.

Reply:
yes! it's something on the top of my priority list.

Reply:
Has the credential-reference been added?

Reply:
@darranl yes!